prost-build/src/lib.rs

```diff
@@ -949,25 +964,32 @@ impl Config {
         outfile.write_all(format!("{}{}\n", ("    ").to_owned().repeat(depth), line).as_bytes())
     }
 
-    fn generate(&mut self, files: Vec<FileDescriptorProto>) -> Result<HashMap<Module, String>> {
+    /// Processes a set of modules and file descriptors, returning a map of modules to generated
+    /// code contents.
+    ///
+    /// This is generally used when control over the output should not be managed by Prost,
+    /// such as in a flow for a `protoc` code generating plugin. When compiling as part of a
+    /// `build.rs` file, instead use [`compile_protos()`].
+    pub fn generate(
+        &mut self,
+        requests: Vec<(Module, FileDescriptorProto)>,
```

Comment:
I think this should accept some generic `impl Iterator` since we don't really care if its a vec or not. What do you think?

Reply:
The issue here is that we need to build up the message graph before we can process each module. That requires stepping through `requests` twice. We can accept an `Iterator` in, but we'd need to stuff it into an intermediate that we can iterate through a second time. Based on my experimentation on the Rust Playground, in release mode, if the input `Iterator` is a `Vec` and it is collected it inside generate, the collection incurs no cost.

We could instead use an iterator that just borrowed the above values (`Item=&(Module, FileDescriptorProto)` or `Item=(&Module, &FileDescriptorProto)`). That would change the return type to `HashMap<&Module, &str>`, but doing so would require cloning the `FileDescriptorProto` (a relatively expensive operation) as `CodeGenerator` wants to consume an owned value (that may be worth looking into changing in the future).

Reply:
Oh that seems like a super valid point, lets just keep the Vec then! Thanks for the thorough explanation!

Reply:
Sounds good. I'll mark this as resolved as well.